### DISCUSSION TITLE
Missing API functions

### DISCUSSION THREAD

**Discussion by elementalTIMING** [elgatosf/elgato#105] [Ideas]
Dear Stream Deck Team,

the Stream Deck is indeed a great product. However, I wonder if it will be further developed. Especially concerning the API, I sorely miss various features. For example these include:

- Creation and navigation of folders. It doesn't seem possible to create a folder with the API and/or navigate through folders.
- I would like to be able to create a plugin (similar to how it works on the Loupedeck) and assign it programmatically a place. Clicking on the plugin would open a folder with already assigned functions. For example, the original Media plugin: here, I have to manually assign each function to a button. As a developer, I wish I could assign a button with my plugin, which acts as a folder. A click would open it, and I could already assign all functions to corresponding places WITHOUT having to manually assign each individual function.
- As enhancement to my previous point: this automatic assignment would also allow dynamically creating/removing buttons. My plugin communicates with a server and receives a list, for example. Unfortunately, with the Stream Deck, I don't seem to have the possibility to automatically create buttons for each received list element. This is possible with the Loupedeck, for example.
Another weakness is that I can't assign an additional action (e.g., a hotkey) to a folder. The detour via profiles is possible but cumbersome, as I can't cleanly realize nested structures here.

Is the API still being developed, and can I "hope" for extensions?

**Comment by ChekTek** [comment]
Hi @elementalPRESS,

Absolutely, we are still expanding the Stream Deck API & SDK. What you have mentioned is technically possible using a profile bundled in your plugin. Essentially, you could create a profile with folders and placeholder actions and you can assign functionality to the placeholders in your profile using the contexts and coordinates you receive in the `willAppear` event. 

https://docs.elgato.com/sdk/plugins/manifest#profiles
https://docs.elgato.com/sdk/plugins/events-received#willappear

While this works, it is very cumbersome to implement. We are actively working on what we are calling "Full Canvas Control". This would allow you to setup an action that takes over the entire Stream Deck canvas (all the keys/dials), which should make what you are describing much easier.

**Reply by elementalTIMING** [reply]
Does it mean that I'll be able to create profiles from within my plugin and/or that I can assign places for buttons from my plugin? Because that would be awesome!

**Reply by ChekTek** [reply]
Keep in mind this is in active development and subject to change, but yes in a sense. It may not provide the full flexibility you are looking for, but you could set your action as a "Full Canvas Control" action and when the user presses the key, you'll receive and event with all the coordinates and id's of the keys available to you at 1 time. You could then take those id's and use them to assign functionality to them.

**Reply by Xanacas** [reply]
Any news on the full canvas control?

**Reply by ChekTek** [reply]
In active development 👍

**Comment by elementalTIMING** [comment]
I started playing with the new cli tool. It created a plugin and the `npm run watch` works wonderful. But I'm wondering how and where I can get console messages? Very simple attempt was to write into the `plugin.ts`

    console.log('Show me');
    console.warn('Show me');
    console.error('Show me');
    console.trace('Show me');

I expected to see the logs on the screen, but I just found the messages in the logs folger. Is there any way to show the logs on the console window too?

Please help!

**Reply by buffos** [reply]
reading through the documentation , probably it was not available on that date, in the [your first changes](https://docs.elgato.com/streamdeck/sdk/introduction/your-first-changes) page in the debugging section it describes how you can see those console.logs
